Additional initializer (#23)
Imageomics/distributed-downloader: src/distributed_downloader/core/initializers/README.md

# Initializers

Initializers are used to process the target dataset into a format that can be used by the downloader.
Spark is used to process the data in parallel.
The following steps are performed by the initializers:

1. Read the dataset from the source.
2. Filter the dataset to remove unwanted data.
3. Rename the columns to match the downloader's schema. Specifically:
    - `identifier` is the url to the image.

>     - `source_id` some identification of the image, that can later be used to access relative information from the
>       source.

Reviewer: ```suggestion
    - `source_id` unique identifier for image using identification from the source; ID that can later be used to access relevant information from the source.
```
For EOL this is redefined later in [`core/mpi_downloader/dataclasses.py`](https://github.com/Imageomics/distributed-downloader/pull/23/files#diff-0ff4264e971b7e4e66120318e5db3e3a1d4c7f4db64a53df932d79703e36c592) to be `<EOL content ID>_<EOL page ID>`--how does this relate if desired in another setting?

The initializers are run only once for each dataset, and the resulting partitioned dataset is used by the downloader to
download the images.

## Structure

### Base Initializer

The `BaseInitializer` class is the base class for all initializers. It contains all the common functionality used by all
initializers.
It has the following methods:

- `load_raw_df`: Reads the dataset from the source.
- `extract_server_name`: Extracts the server name from the `identifier`.
- `generate_uuid`: Generates a unique identifier for each image.
- `partition_dataframe`: Partitions the dataset first by the server name and then into smaller partitions with a fixed
  size.
- `save_results`: Saves the partitioned dataset to the target location in `parquet` format.

### Initializers

The initializers are classes that inherit from the `BaseInitializer` class. They implement the specific logic for each
dataset.
The following initializers are available:

- `GBIFInitializer`: Initializer for the GBIF dataset. It filters out any entries without an `gbifID` or `identifier`
  value.
  Additionally, removes any entries that are not `StillImage` by type or `image` by format.

>   And lastly, it removes any entries that have `MATERIAL_CITATION` in `basisOfRecord`.

Reviewer: ```suggestion
  And lastly, it removes any entries that have `MATERIAL_CITATION` in `basisOfRecord`. This is because these are known to be images of text documents.
```

- `FathomNetInitializer`: Initializer for the FathomNet dataset. It filters out any entries without an `uuid` or `url`
  value.

>   Additionally, removes any entries that are "not valid" by the `valid` column.

Reviewer: Do you recall what determined if an entry was `valid` in FathomNet?

> - `EoLInitializer`: Initializer for the EoL dataset. It filters out any entries without an `EOL content ID` or

Reviewer: ```suggestion
- `EoLInitializer`: Initializer for the EOL dataset. It filters out any entries without an `EOL content ID` or
```

>   `EOL Full-Size Copy URL` value.

Reviewer: ```suggestion
  `EOL Full-Size Copy URL` value.
    - The `EOL content ID` is set as the `source_id`, which is used to map to the original metadata file to get the `EOL page ID` to match to the `taxon.tab` for taxa information. `EOL content ID` is not a persistent identifier at EOL, so it is important to maintain the original metadata file.
```

> - `LilaInitializer`: Initializer for the Lila dataset. It filters out any entries that do not have a `url` value (by

Reviewer: ```suggestion
- `LilaInitializer`: Initializer for the LILA dataset. It filters out any entries that do not have a `url` value (by
```

  `url_gcp`, `url_aws` or `url_azure`).

>   Additionally, removes any entries that are `emptu` by the `original_label` column.

Reviewer: ```suggestion
  Additionally, removes any entries that are `empty` by the `original_label` column.
```

# Initializers

Initializers are used to process the target dataset into a format that can be used by the downloader.
Spark is used to process the data in parallel.
The following steps are performed by the initializers:

1. Read the dataset from the source.
2. Filter the dataset to remove unwanted data.
3. Rename the columns to match the downloader's schema. Specifically:
    - `identifier` is the url to the image.

>     - `source_id` some identification of the image, that can later be used to access relative information from the

Reviewer: ```suggestion
    - `source_id` some identification of the image, that can later be used to access relevant information from the
```
